src/main/scala-2.11/markets/clearing/engines/CDAMatchingEngine.scala

```diff
@@ -15,24 +15,208 @@ limitations under the License.
 */
 package markets.clearing.engines
 
-import java.util.UUID
-
-import markets.clearing.strategies.BestLimitPriceFormationStrategy
-import markets.orders.orderings.PriceOrdering
+import markets.clearing.engines.matches.{Match, PartialMatch, TotalMatch}
+import markets.orders.limit.{LimitBidOrder, LimitAskOrder, LimitOrderLike}
+import markets.orders.market.{MarketOrderLike, MarketAskOrder, MarketBidOrder}
 import markets.orders.{AskOrder, BidOrder, Order}
+import markets.orders.orderings.PriceOrdering
+
+import scala.annotation.tailrec
+import scala.collection.{immutable, mutable}
 
-import scala.collection.immutable
 
+/** Continuous Double Auction (CDA) Matching Engine. */
+class CDAMatchingEngine(askOrdering: PriceOrdering[AskOrder],
+                        bidOrdering: PriceOrdering[BidOrder],
+                        initialPrice: Long) extends MutableMatchingEngine {
 
-class CDAMatchingEngine(val askOrdering: PriceOrdering[AskOrder],
-                        val bidOrdering: PriceOrdering[BidOrder],
-                        initialPrice: Long)
-  extends CDAMatchingEngineLike
-  with BestLimitPriceFormationStrategy {
+  /* Mutable collection of ask orders for internal use only! */
+  protected val _askOrderBook = mutable.TreeSet.empty[AskOrder](askOrdering)
 
-  protected var orderBook = immutable.Set.empty[Order]
+  /* Mutable collection of bid orders for internal use only! */
+  protected val _bidOrderBook = mutable.TreeSet.empty[BidOrder](bidOrdering)
 
-  protected var mostRecentPrice = initialPrice
+  /* Cached value of most recent transaction price for internal use only. */
+  private var _mostRecentPrice = initialPrice
+
+  protected def crosses(incomingOrder: Order, existingOrder: Order): Boolean = {
+    (incomingOrder, existingOrder) match {
+      case (ask: AskOrder, bid: BidOrder) => ask.price <= bid.price
+      case (bid: BidOrder, ask: AskOrder) => bid.price >= ask.price
+    }
+  }
+
+  /** A sorted collection of ask orders.
+    * @note This is an immutable view into a mutable private collection.
+    */
+  def askOrderBook: immutable.TreeSet[AskOrder] = {
+    immutable.TreeSet.empty[AskOrder](askOrdering) ++ _askOrderBook
+  }
+
+  /** A sorted collection of bid orders.
+    * @note This is an immutable view into a mutable private collection.
+    */
+  def bidOrderBook: immutable.TreeSet[BidOrder] = {
+    immutable.TreeSet.empty[BidOrder](bidOrdering) ++ _bidOrderBook
+  }
+
+  def findMatch(incomingOrder: Order): Option[immutable.Iterable[Match]] = {
+
+    incomingOrder match {
```

Comment:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Default case not specified.](https://www.codacy.com/app/davidrpugh/markets-sandbox/file/1519375750/issues/source?bid=2921361&fileBranchId=3008855#l65)